staging/src/k8s.io/client-go/tools/cache/controller.go

```diff
@@ -585,6 +606,95 @@ func processDeltas(
 	return nil
 }
 
+// processDeltasInBatch applies a batch of Deltas to the given Store and
+// notifies the ResourceEventHandler of add, update, or delete events.
+//
+// If the Store supports transactions (TransactionStore), all Deltas are applied
+// atomically in a single transaction and corresponding handler callbacks are
+// executed afterward. Otherwise, each Delta is processed individually.
+//
+// Returns an error if any Delta or transaction fails. For TransactionError,
+// only successful operations trigger callbacks.
+func processDeltasInBatch(
+	handler ResourceEventHandler,
+	clientState Store,
+	deltasList []Deltas,
+	isInInitialList bool,
+) error {
+	// from oldest to newest
+	txns := make([]Transaction, 0)
+	callbacks := make([]func(), 0)
+	txnStore, txnSupported := clientState.(TransactionStore)
+	if !txnSupported {
+		var errs []error
+		for _, delta := range deltasList {
+			if err := processDeltas(handler, clientState, delta, isInInitialList); err != nil {
```

Comment:
only when falling back do we need to wrap each Delta to Deltas (just like the old non-batch Pop path did):
```suggestion
			if err := processDeltas(handler, clientState, Deltas{delta}, isInInitialList); err != nil {
```

Reply:
refactored